File: backend/app/models/survey.py
"""
Survey models for A/B blind evaluation of ParliamentRAG vs Baseline RAG,
and simple Likert-scale evaluation for queries without a predefined baseline.

A/B evaluation: users rate both responses (blind) and indicate preference.
Simple evaluation: users rate only the system response on 4 Likert dimensions.
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Literal
from datetime import datetime
from uuid import uuid4
from enum import Enum


class RatingScale(int, Enum):
    """1-5 star rating scale"""
    VERY_POOR = 1
    POOR = 2
    AVERAGE = 3
    GOOD = 4
    EXCELLENT = 5


# Valid issue tags for individual citation evaluation
CITATION_ISSUES = [
    "out_of_context",
    "truncated",
    "wrong_attribution",
    "duplicate",
    "unverifiable",
    "none",
]


class CitationEvaluation(BaseModel):
    """Evaluation of a single citation within a response."""
    evidence_id: str = Field(..., description="chunk_id of the citation")
    relevance: int = Field(..., ge=1, le=5, description="How relevant is this citation to the query/response")
    faithfulness: int = Field(..., ge=1, le=5, description="How faithfully does the quote represent the original speech")
    informativeness: int = Field(..., ge=1, le=5, description="How much value does this citation add")
    attribution: Literal["correct", "incorrect", "unverifiable"] = Field(
        ..., description="Is the speaker/group/date attribution correct"
    )
    issues: List[str] = Field(default_factory=list, description="Issue tags for quick triage")


class ABRating(BaseModel):
    """Rating for a single dimension, comparing Response A and Response B."""
    rating_a: int = Field(..., ge=1, le=5, description="Rating for Risposta A")
    rating_b: int = Field(..., ge=1, le=5, description="Rating for Risposta B")
    preference: Literal["A", "B", "equal"] = Field(..., description="Which response is better")


class SurveyResponse(BaseModel):
    """Complete A/B survey response for a single chat session."""
    id: str = Field(default_factory=lambda: str(uuid4()))
    chat_id: str = Field(..., description="Associated chat history ID")
    timestamp: datetime = Field(default_factory=datetime.now)

    # A/B comparative ratings per dimension
    answer_quality: ABRating
    answer_clarity: ABRating
    answer_completeness: ABRating
    citations_relevance: ABRating
    citations_accuracy: ABRating
    balance_perception: ABRating
    balance_fairness: ABRating

    # Overall
    overall_satisfaction_a: int = Field(..., ge=1, le=5, description="Overall satisfaction for Risposta A")
    overall_satisfaction_b: int = Field(..., ge=1, le=5, description="Overall satisfaction for Risposta B")
    overall_preference: Literal["A", "B", "equal"] = Field(..., description="Overall preferred response")

    would_recommend: bool = Field(..., description="Would recommend to colleagues")

    # Qualitative feedback
    feedback_positive: Optional[str] = Field(None, max_length=1000, description="What worked well")
    feedback_improvement: Optional[str] = Field(None, max_length=1000, description="Suggestions for improvement")

    # Individual citation evaluations
    citation_evaluations_a: List[CitationEvaluation] = Field(default_factory=list, description="Per-citation evaluations for Response A")
    citation_evaluations_b: List[CitationEvaluation] = Field(default_factory=list, description="Per-citation evaluations for Response B")

    # Source authority evaluation — optional (added after initial schema; None for legacy records)
    source_relevance: Optional[ABRating] = Field(None, description="Rilevanza tematica degli esperti citati")
    source_authority: Optional[ABRating] = Field(None, description="Autorevolezza istituzionale degli esperti")
    source_coverage: Optional[ABRating] = Field(None, description="Copertura delle coalizioni")

    # Metadata
    evaluator_role: Optional[str] = Field(None, description="Role of evaluator")
    evaluation_context: Optional[str] = Field(None, description="Context of evaluation")
    evaluator_id: Optional[str] = Field(None, description="Unique identifier for the evaluator")
    ab_assignment: Optional[Dict[str, str]] = Field(None, description="Per-evaluator A/B assignment, e.g. {'A': 'system', 'B': 'baseline'}")


class SurveyResponseCreate(BaseModel):
    """Model for creating a new A/B survey response."""
    chat_id: str

    # A/B comparative ratings per dimension
    answer_quality: ABRating
    answer_clarity: ABRating
    answer_completeness: ABRating
    citations_relevance: ABRating
    citations_accuracy: ABRating
    balance_perception: ABRating
    balance_fairness: ABRating

    # Overall
    overall_satisfaction_a: int = Field(..., ge=1, le=5)
    overall_satisfaction_b: int = Field(..., ge=1, le=5)
    overall_preference: Literal["A", "B", "equal"]

    would_recommend: bool

    feedback_positive: Optional[str] = None
    feedback_improvement: Optional[str] = None

    # Individual citation evaluations
    citation_evaluations_a: List[CitationEvaluation] = Field(default_factory=list)
    citation_evaluations_b: List[CitationEvaluation] = Field(default_factory=list)

    # Source authority evaluation (optional)
    source_relevance: Optional[ABRating] = None
    source_authority: Optional[ABRating] = None
    source_coverage: Optional[ABRating] = None

    evaluator_role: Optional[str] = None
    evaluation_context: Optional[str] = None

    # Evaluator identifier (for multi-evaluator study)
    evaluator_id: Optional[str] = Field(None, description="Unique identifier for the evaluator (e.g. from ?evaluator= URL param)")

    # A/B assignment (provided by frontend when using evaluation_set baseline)
    ab_assignment: Optional[Dict[str, str]] = Field(
        None, description="e.g. {'A': 'system', 'B': 'baseline'} — set if using evaluation_set baseline"
    )
    evaluation_set_topic: Optional[str] = Field(
        None, description="Topic name from evaluation_set.json if this is an evaluation_set A/B"
    )


# ── Simple Likert Evaluation (for queries without a predefined baseline) ──────

# Dimensions for simple Likert evaluation
SIMPLE_DIMENSIONS = [
    "answer_clarity",
    "answer_quality",
    "balance_perception",
    "balance_fairness",
]

SIMPLE_DIMENSION_LABELS = {
    "answer_clarity": "Chiarezza espositiva",
    "answer_quality": "Qualità complessiva percepita",
    "balance_perception": "Bilanciamento percepito",
    "balance_fairness": "Equità rappresentazione",
}


class SimpleRatingResponse(BaseModel):
    """Simple Likert-scale evaluation of a single chat response (no baseline needed)."""
    id: str = Field(default_factory=lambda: str(uuid4()))
    chat_id: str = Field(..., description="Associated chat history ID")
    timestamp: datetime = Field(default_factory=datetime.now)

    answer_clarity: int = Field(..., ge=1, le=5, description="Chiarezza espositiva (1-5)")
    answer_quality: int = Field(..., ge=1, le=5, description="Qualità complessiva percepita (1-5)")
    balance_perception: int = Field(..., ge=1, le=5, description="Bilanciamento percepito (1-5)")
    balance_fairness: int = Field(..., ge=1, le=5, description="Equità rappresentazione (1-5)")

    feedback: Optional[str] = Field(None, max_length=1000, description="Commento libero opzionale")
    evaluator_id: Optional[str] = Field(None, description="Unique identifier for the evaluator")


class SimpleRatingCreate(BaseModel):
    """Model for creating a new simple Likert rating."""
    chat_id: str
    answer_clarity: int = Field(..., ge=1, le=5)
    answer_quality: int = Field(..., ge=1, le=5)
    balance_perception: int = Field(..., ge=1, le=5)
    balance_fairness: int = Field(..., ge=1, le=5)
    feedback: Optional[str] = None
    evaluator_id: Optional[str] = Field(None, description="Unique identifier for the evaluator")


class SurveyWithChat(BaseModel):
    """Survey response with associated chat metadata."""
    survey: SurveyResponse
    chat_query: str
    chat_preview: str
    chat_timestamp: datetime


# Dimensions used for A/B comparison
AB_DIMENSIONS = [
    "answer_quality",
    "answer_clarity",
    "answer_completeness",
    "citations_relevance",
    "citations_accuracy",
    "balance_perception",
    "balance_fairness",
    "source_relevance",
    "source_authority",
    "source_coverage",
]

# Optional dims added after the initial schema (may be absent in legacy Neo4j records)
OPTIONAL_AB_DIMS = {"source_relevance", "source_authority", "source_coverage"}


class SurveyStats(BaseModel):
    """Aggregated A/B survey statistics (de-blinded)."""
    total_surveys: int

    # De-blinded averages per dimension
    system_avg_per_dimension: Dict[str, float]
    baseline_avg_per_dimension: Dict[str, float]

    # Overall satisfaction averages
    system_avg_overall: float
    baseline_avg_overall: float

    # Win rates (de-blinded)
    system_win_rate: float   # % of preferences favoring system
    baseline_win_rate: float  # % of preferences favoring baseline
    tie_rate: float           # % equal

    # Per-dimension preference counts (de-blinded)
    per_dimension_preference: Dict[str, Dict[str, int]]  # dim -> {"system": N, "baseline": N, "equal": N}

    recommendation_rate: float


class SurveyListResponse(BaseModel):
    """Response for listing surveys."""
    surveys: List[SurveyWithChat]
    total: int
    stats: Optional[SurveyStats] = None


# Survey questions configuration (for frontend) - A/B format
SURVEY_QUESTIONS = [
    {
        "id": "answer_quality",
        "category": "Qualita Risposta",
        "question": "Qualita complessiva della risposta",
        "description": "Considera l'utilita pratica per il lavoro giornalistico"
    },
    {
        "id": "answer_clarity",
        "category": "Qualita Risposta",
        "question": "Chiarezza e leggibilita della risposta",
        "description": "Valuta la struttura, il linguaggio e la facilita di comprensione"
    },
    {
        "id": "answer_completeness",
        "category": "Qualita Risposta",
        "question": "Completezza delle informazioni",
        "description": "Considera se mancano informazioni importanti"
    },
    {
        "id": "citations_relevance",
        "category": "Citazioni",
        "question": "Pertinenza delle citazioni parlamentari",
        "description": "Valuta se le citazioni supportano effettivamente la risposta"
    },
    {
        "id": "citations_accuracy",
        "category": "Citazioni",
        "question": "Accuratezza delle attribuzioni",
        "description": "Considera se deputato, data e contesto sono corretti"
    },
    {
        "id": "balance_perception",
        "category": "Bilanciamento Politico",
        "question": "Bilanciamento politico percepito",
        "description": "Valuta se sono rappresentate diverse posizioni politiche"
    },
    {
        "id": "balance_fairness",
        "category": "Bilanciamento Politico",
        "question": "Equita nella rappresentazione",
        "description": "Considera se c'e imparzialita nella presentazione"
    },
    {
        "id": "source_relevance",
        "category": "Autorità Esperti",
        "question": "Rilevanza tematica degli esperti",
        "description": "Gli esperti citati sono le voci parlamentari più attive e pertinenti su questo tema?"
    },
    {
        "id": "source_authority",
        "category": "Autorità Esperti",
        "question": "Autorevolezza istituzionale",
        "description": "Le figure citate ricoprono ruoli istituzionali rilevanti (commissioni, presidenze, portavoce)?"
    },
    {
        "id": "source_coverage",
        "category": "Autorità Esperti",
        "question": "Copertura delle coalizioni",
        "description": "Le principali forze politiche sono rappresentate in modo equilibrato?"
    },
    {
        "id": "overall_satisfaction",
        "category": "Valutazione Complessiva",
        "question": "Soddisfazione complessiva",
        "description": "Valutazione generale dell'esperienza con ciascuna risposta"
    }
]

File: frontend/src/types/survey.ts
/**
 * Survey types for A/B blind evaluation of ParliamentRAG vs Baseline RAG.
 */

export interface SurveyQuestion {
  id: string;
  category: string;
  question: string;
  description: string;
}

export interface ABRating {
  rating_a: number;
  rating_b: number;
  preference: "A" | "B" | "equal" | "";
}

export interface CitationEvaluation {
  evidence_id: string;
  relevance: number;      // 1-5
  faithfulness: number;    // 1-5
  informativeness: number; // 1-5
  attribution: "correct" | "incorrect" | "unverifiable" | "";
  issues: string[];
}

export const CITATION_ISSUES = [
  { id: "none", label: "Nessun problema", icon: "check" },
  { id: "out_of_context", label: "Fuori contesto", icon: "alert" },
  { id: "truncated", label: "Citazione troncata", icon: "scissors" },
  { id: "wrong_attribution", label: "Attribuzione errata", icon: "user-x" },
  { id: "duplicate", label: "Duplicata", icon: "copy" },
  { id: "unverifiable", label: "Non verificabile", icon: "help" },
] as const;

export const getInitialCitationEvaluation = (evidenceId: string): CitationEvaluation => ({
  evidence_id: evidenceId,
  relevance: 0,
  faithfulness: 0,
  informativeness: 0,
  attribution: "",
  issues: [],
});

export interface SurveyResponse {
  id: string;
  chat_id: string;
  timestamp: string;

  // A/B comparative ratings per dimension
  answer_quality: ABRating;
  answer_clarity: ABRating;
  answer_completeness: ABRating;
  citations_relevance: ABRating;
  citations_accuracy: ABRating;
  balance_perception: ABRating;
  balance_fairness: ABRating;

  // Overall
  overall_satisfaction_a: number;
  overall_satisfaction_b: number;
  overall_preference: "A" | "B" | "equal";

  would_recommend: boolean;

  // Qualitative feedback
  feedback_positive?: string;
  feedback_improvement?: string;

  // Individual citation evaluations
  citation_evaluations_a: CitationEvaluation[];
  citation_evaluations_b: CitationEvaluation[];

  // Metadata
  evaluator_role?: string;
  evaluation_context?: string;
  evaluator_id?: string;
}

export interface SurveyResponseCreate {
  chat_id: string;

  answer_quality: ABRating;
  answer_clarity: ABRating;
  answer_completeness: ABRating;
  citations_relevance: ABRating;
  citations_accuracy: ABRating;
  balance_perception: ABRating;
  balance_fairness: ABRating;
  source_relevance: ABRating;
  source_authority: ABRating;
  source_coverage: ABRating;

  overall_satisfaction_a: number;
  overall_satisfaction_b: number;
  overall_preference: "A" | "B" | "equal";

  would_recommend: boolean;

  feedback_positive?: string;
  feedback_improvement?: string;

  citation_evaluations_a?: CitationEvaluation[];
  citation_evaluations_b?: CitationEvaluation[];

  evaluator_role?: string;
  evaluation_context?: string;
  evaluator_id?: string;

  // For evaluation_set A/B flow: frontend-generated assignment stored on ChatHistory
  ab_assignment?: Record<string, string>; // e.g. { A: "system", B: "baseline" }
  evaluation_set_topic?: string;

  // Baseline authority: mean authority score of experts found in baseline response
  baseline_authority_avg?: number;

  // Per-group authority votes: group_key → -1 (A better), 0 (equal/same), 1 (B better)
  group_authority_votes?: Record<string, number>;
}

export interface SurveyWithChat {
  survey: SurveyResponse;
  chat_query: string;
  chat_preview: string;
  chat_timestamp: string;
}

export interface SurveyStats {
  total_surveys: number;
  system_avg_per_dimension: Record<string, number>;
  baseline_avg_per_dimension: Record<string, number>;
  system_avg_overall: number;
  baseline_avg_overall: number;
  system_win_rate: number;
  baseline_win_rate: number;
  tie_rate: number;
  per_dimension_preference: Record<string, Record<string, number>>;
  recommendation_rate: number;
}

export interface SurveyListResponse {
  surveys: SurveyWithChat[];
  total: number;
  stats?: SurveyStats;
}

export interface SimpleRatingResponse {
  id: string;
  chat_id: string;
  timestamp: string;
  answer_clarity: number;
  answer_quality: number;
  balance_perception: number;
  balance_fairness: number;
  feedback?: string;
  evaluator_id?: string;
}

export interface SimpleRatingCreate {
  chat_id: string;
  answer_clarity: number;
  answer_quality: number;
  balance_perception: number;
  balance_fairness: number;
  feedback?: string;
  evaluator_id?: string;
}

export const SIMPLE_DIMENSIONS = [
  "answer_clarity",
  "answer_quality",
  "balance_perception",
  "balance_fairness",
] as const;

export type SimpleDimension = typeof SIMPLE_DIMENSIONS[number];

export const SIMPLE_DIMENSION_LABELS: Record<SimpleDimension, string> = {
  answer_clarity: "Chiarezza espositiva",
  answer_quality: "Qualita complessiva percepita",
  balance_perception: "Bilanciamento percepito",
  balance_fairness: "Equita rappresentazione",
};

export interface SimpleRatingFormState {
  answer_clarity: number;
  answer_quality: number;
  balance_perception: number;
  balance_fairness: number;
  feedback: string;
}

export const getInitialSimpleRatingFormState = (): SimpleRatingFormState => ({
  answer_clarity: 0,
  answer_quality: 0,
  balance_perception: 0,
  balance_fairness: 0,
  feedback: "",
});

export interface BaselineCitation {
  speaker: string;
  group: string;
  date: string;
  quote: string;
  verbatim_verified: boolean;
}

export interface PendingChat {
  id: string;
  query: string;
  preview: string;
  timestamp: string;
  evaluation_type: "ab" | "simple";
  matched_topic?: string;
  baseline_answer?: string;
  baseline_experts?: any[];
  baseline_citations?: BaselineCitation[];
}

export interface PendingChatsResponse {
  pending: PendingChat[];
  total: number;
}

// A/B dimensions for comparison
export const AB_DIMENSIONS = [
  "answer_quality",
  "answer_clarity",
  "answer_completeness",
  "citations_relevance",
  "citations_accuracy",
  "balance_perception",
  "balance_fairness",
  "source_relevance",
  "source_authority",
  "source_coverage",
] as const;

export type ABDimension = typeof AB_DIMENSIONS[number];

// Survey form state for A/B evaluation
export interface SurveyFormState {
  answer_quality: ABRating;
  answer_clarity: ABRating;
  answer_completeness: ABRating;
  citations_relevance: ABRating;
  citations_accuracy: ABRating;
  balance_perception: ABRating;
  balance_fairness: ABRating;
  // Source authority dimensions (grounded in TREC Expert Finding + DQI literature)
  source_relevance: ABRating;
  source_authority: ABRating;
  source_coverage: ABRating;
  overall_satisfaction_a: number;
  overall_satisfaction_b: number;
  overall_preference: "A" | "B" | "equal" | "";
  would_recommend: boolean;
  feedback_positive: string;
  feedback_improvement: string;
  citation_evaluations_a: CitationEvaluation[];
  citation_evaluations_b: CitationEvaluation[];
}

// Survey questions configuration - A/B format
export const SURVEY_QUESTIONS: SurveyQuestion[] = [
  {
    id: "answer_quality",
    category: "Qualita Risposta",
    question: "Qualita complessiva della risposta",
    description: "Considera l'utilita pratica per il lavoro giornalistico",
  },
  {
    id: "answer_clarity",
    category: "Qualita Risposta",
    question: "Chiarezza e leggibilita della risposta",
    description: "Valuta la struttura, il linguaggio e la facilita di comprensione",
  },
  {
    id: "answer_completeness",
    category: "Qualita Risposta",
    question: "Completezza delle informazioni",
    description: "Considera se mancano informazioni importanti",
  },
  {
    id: "citations_relevance",
    category: "Citazioni",
    question: "Pertinenza delle citazioni parlamentari",
    description: "Valuta se le citazioni supportano effettivamente la risposta",
  },
  {
    id: "citations_accuracy",
    category: "Citazioni",
    question: "Accuratezza delle attribuzioni",
    description: "Considera se deputato, data e contesto sono corretti",
  },
  {
    id: "balance_perception",
    category: "Bilanciamento Politico",
    question: "Bilanciamento politico percepito",
    description: "Valuta se sono rappresentate diverse posizioni politiche",
  },
  {
    id: "balance_fairness",
    category: "Bilanciamento Politico",
    question: "Equita nella rappresentazione",
    description: "Considera se c'e imparzialita nella presentazione",
  },
  // Authority dimensions — Autorità Esperti category
  // Based on: TREC Expert Finding (Craswell et al. 2005), DQI (Steenbergen et al. 2003),
  // Source Credibility Theory (Hovland et al. 1953; Gaziano & McGrath 1986)
  {
    id: "source_relevance",
    category: "Autorità Esperti",
    question: "Rilevanza tematica degli esperti",
    description: "Gli esperti citati sono le voci parlamentari più attive e pertinenti su questo tema?",
  },
  {
    id: "source_authority",
    category: "Autorità Esperti",
    question: "Autorevolezza istituzionale",
    description: "Le figure citate ricoprono ruoli rilevanti (presidenze di commissione, portavoce, governo)?",
  },
  {
    id: "source_coverage",
    category: "Autorità Esperti",
    question: "Copertura delle coalizioni",
    description: "Le principali forze politiche sono rappresentate in modo equilibrato tra maggioranza e opposizione?",
  },
  {
    id: "overall_satisfaction",
    category: "Valutazione Complessiva",
    question: "Soddisfazione complessiva",
    description: "Valutazione generale dell'esperienza con ciascuna risposta",
  },
];

// Initial form state
export const getInitialABRating = (): ABRating => ({
  rating_a: 0,
  rating_b: 0,
  preference: "",
});

export const getInitialSurveyFormState = (): SurveyFormState => ({
  answer_quality: getInitialABRating(),
  answer_clarity: getInitialABRating(),
  answer_completeness: getInitialABRating(),
  citations_relevance: getInitialABRating(),
  citations_accuracy: getInitialABRating(),
  balance_perception: getInitialABRating(),
  balance_fairness: getInitialABRating(),
  source_relevance: getInitialABRating(),
  source_authority: getInitialABRating(),
  source_coverage: getInitialABRating(),
  overall_satisfaction_a: 0,
  overall_satisfaction_b: 0,
  overall_preference: "",
  would_recommend: false,
  feedback_positive: "",
  feedback_improvement: "",
  citation_evaluations_a: [],
  citation_evaluations_b: [],
});
